In [1]:
import numpy as np
import pandas as pd

---

### 1. Loading and Exploring Data
Read the ifood.xlsx file using Pandas and store it in the variable df.

**คำอธิบาย:** ข้อนี้เป็นขั้นตอนแรกสุดของการทำงานกับข้อมูล คือการอ่านไฟล์ Excel (.xlsx) เข้ามาเก็บไว้ในตัวแปร โดยใช้คำสั่ง `pd.read_excel("ifood.xlsx")` จากไลบรารี pandas (ที่ import มาเป็นชื่อย่อ `pd` ในเซลล์แรก) ฟังก์ชันนี้จะอ่านข้อมูลจากไฟล์ Excel แล้วแปลงเป็นโครงสร้างข้อมูลแบบตาราง 2 มิติที่เรียกว่า **DataFrame** ซึ่งมีทั้งแถว (records/ลูกค้าแต่ละคน) และคอลัมน์ (attributes/คุณสมบัติต่าง ๆ เช่น รายได้ อายุ การศึกษา)

ผลลัพธ์ถูกเก็บไว้ในตัวแปรชื่อ `df` (ย่อมาจาก DataFrame) ซึ่งจะถูกนำไปใช้ต่อในแบบฝึกหัดข้อถัด ๆ ไปทั้งหมด เมื่อพิมพ์ `df` ต่อท้ายจะแสดงตัวอย่างข้อมูลบางส่วน (แถวแรกและแถวท้ายบางส่วน) พร้อมบอกจำนวนแถวและคอลัมน์ทั้งหมดของข้อมูล (2205 แถว, 27 คอลัมน์)

In [2]:
df = pd.read_excel("ifood.xlsx")

df

,No,Income,Kidhome,Teenhome,Recency,MntWines,MntFruits,MntMeatProducts,MntFishProducts,MntSweetProducts,...,AcceptedCmp4,AcceptedCmp5,AcceptedCmp1,AcceptedCmp2,Complain,Response,Age,Customer_Days,Marital,Education
0,1,58138,0,0,58,635,88,546,172,88,...,0,0,0,0,0,1,63,2822,Single,Graduation
1,2,46344,1,1,38,11,1,6,2,1,...,0,0,0,0,0,0,66,2272,Single,Graduation
2,3,71613,0,0,26,426,49,127,111,21,...,0,0,0,0,0,0,55,2471,Together,Graduation
3,4,26646,1,0,26,11,4,20,10,3,...,0,0,0,0,0,0,36,2298,Together,Graduation
4,5,58293,1,0,94,173,43,118,46,27,...,0,0,0,0,0,0,39,2320,Married,PhD
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
2200,2201,61223,0,1,46,709,43,182,42,118,...,0,0,0,0,0,0,53,2540,Married,Graduation
2201,2202,64014,2,1,56,406,0,30,0,0,...,0,0,1,0,0,0,74,2178,Together,PhD
2202,2203,56981,0,0,91,908,48,217,32,12,...,1,0,0,0,0,0,39,2314,Divorced,Graduation
2203,2204,69245,0,1,8,428,30,214,80,30,...,0,0,0,0,0,0,64,2315,Together,Master


---

### 2. Inspect the First Records
Display the first 5 records and the first 10 records of the dataset.

**คำอธิบาย:** ข้อนี้ใช้เมธอด `df.head()` เพื่อดูข้อมูลแถวแรก ๆ ของ DataFrame แบบรวดเร็ว โดย default ของ `.head()` (ไม่ใส่ค่าใด ๆ ในวงเล็บ) จะแสดง 5 แถวแรก และหากต้องการจำนวนแถวอื่นก็ระบุตัวเลขในวงเล็บได้ เช่น `df.head(10)` เพื่อแสดง 10 แถวแรก

การดูข้อมูลตัวอย่างแบบนี้เป็นขั้นตอนแรก ๆ ที่ควรทำเสมอหลังโหลดข้อมูล เพื่อตรวจสอบเบื้องต้นว่าข้อมูลถูกอ่านเข้ามาถูกต้อง มีคอลัมน์และรูปแบบค่าตรงตามที่คาดไว้หรือไม่ ก่อนที่จะไปวิเคราะห์ข้อมูลทั้งหมด

In [3]:
### First 5 records ###
df.head()

,No,Income,Kidhome,Teenhome,Recency,MntWines,MntFruits,MntMeatProducts,MntFishProducts,MntSweetProducts,...,AcceptedCmp4,AcceptedCmp5,AcceptedCmp1,AcceptedCmp2,Complain,Response,Age,Customer_Days,Marital,Education
0,1,58138,0,0,58,635,88,546,172,88,...,0,0,0,0,0,1,63,2822,Single,Graduation
1,2,46344,1,1,38,11,1,6,2,1,...,0,0,0,0,0,0,66,2272,Single,Graduation
2,3,71613,0,0,26,426,49,127,111,21,...,0,0,0,0,0,0,55,2471,Together,Graduation
3,4,26646,1,0,26,11,4,20,10,3,...,0,0,0,0,0,0,36,2298,Together,Graduation
4,5,58293,1,0,94,173,43,118,46,27,...,0,0,0,0,0,0,39,2320,Married,PhD


In [4]:
### First 10 records ###
df.head(10)

,No,Income,Kidhome,Teenhome,Recency,MntWines,MntFruits,MntMeatProducts,MntFishProducts,MntSweetProducts,...,AcceptedCmp4,AcceptedCmp5,AcceptedCmp1,AcceptedCmp2,Complain,Response,Age,Customer_Days,Marital,Education
0,1,58138,0,0,58,635,88,546,172,88,...,0,0,0,0,0,1,63,2822,Single,Graduation
1,2,46344,1,1,38,11,1,6,2,1,...,0,0,0,0,0,0,66,2272,Single,Graduation
2,3,71613,0,0,26,426,49,127,111,21,...,0,0,0,0,0,0,55,2471,Together,Graduation
3,4,26646,1,0,26,11,4,20,10,3,...,0,0,0,0,0,0,36,2298,Together,Graduation
4,5,58293,1,0,94,173,43,118,46,27,...,0,0,0,0,0,0,39,2320,Married,PhD
5,6,62513,0,1,16,520,42,98,0,42,...,0,0,0,0,0,0,53,2452,Together,Master
6,7,55635,0,1,34,235,65,164,50,49,...,0,0,0,0,0,0,49,2752,Divorced,Graduation
7,8,33454,1,0,32,76,10,56,3,1,...,0,0,0,0,0,0,35,2576,Married,PhD
8,9,30351,1,0,19,14,0,24,3,3,...,0,0,0,0,0,1,46,2547,Together,PhD
9,10,5648,1,1,68,28,0,6,1,1,...,0,0,0,0,0,0,70,2267,Together,PhD


---

### 3. Inspect the Last Records
Display the last 5 records of the dataset.

**คำอธิบาย:** ข้อนี้ใช้เมธอด `df.tail()` เพื่อดูข้อมูล 5 แถวสุดท้ายของ DataFrame (ตรงข้ามกับ `.head()` ที่ดูแถวแรก) โดย default ของ `.tail()` จะแสดง 5 แถว หากต้องการจำนวนอื่นสามารถระบุตัวเลขในวงเล็บได้ เช่น `df.tail(10)`

การตรวจสอบแถวสุดท้ายมีประโยชน์ในการยืนยันว่าข้อมูลถูกโหลดมาครบถ้วนจนถึงบรรทัดสุดท้ายจริง ไม่มีการตัดข้อมูลขาดหายไปในระหว่างการอ่านไฟล์ และช่วยให้เห็นภาพรวมของข้อมูลทั้งต้นและท้ายไฟล์

In [5]:
df.tail()

,No,Income,Kidhome,Teenhome,Recency,MntWines,MntFruits,MntMeatProducts,MntFishProducts,MntSweetProducts,...,AcceptedCmp4,AcceptedCmp5,AcceptedCmp1,AcceptedCmp2,Complain,Response,Age,Customer_Days,Marital,Education
2200,2201,61223,0,1,46,709,43,182,42,118,...,0,0,0,0,0,0,53,2540,Married,Graduation
2201,2202,64014,2,1,56,406,0,30,0,0,...,0,0,1,0,0,0,74,2178,Together,PhD
2202,2203,56981,0,0,91,908,48,217,32,12,...,1,0,0,0,0,0,39,2314,Divorced,Graduation
2203,2204,69245,0,1,8,428,30,214,80,30,...,0,0,0,0,0,0,64,2315,Together,Master
2204,2205,52869,1,1,40,84,3,61,2,1,...,0,0,0,0,0,1,66,2781,Married,PhD


---

### 4. Inspect Dataset Information
Check the following: (1) Number of records, (2) Number of columns, (3) Data type of each column, (4) Number of non-null values

**คำอธิบาย:** ข้อนี้ต้องการภาพรวมโครงสร้างของ DataFrame โดยแสดงได้ 2 วิธี:
1. **ตรวจสอบทีละรายการ** - ใช้ `df.shape[0]` เพื่อดูจำนวนแถว (records), `df.shape[1]` เพื่อดูจำนวนคอลัมน์, `df.dtypes` เพื่อดูชนิดข้อมูลของแต่ละคอลัมน์ (เช่น int64 สำหรับตัวเลข, object/str สำหรับข้อความ) และ `df.notnull().sum()` เพื่อนับจำนวนค่าที่ "ไม่ใช่ค่าว่าง" (non-null) ในแต่ละคอลัมน์
2. **ดูทั้งหมดในคำสั่งเดียว** - ใช้ `df.info()` ซึ่งสรุปทั้ง 4 อย่างข้างต้นไว้ในรายงานเดียว ได้แก่ จำนวนแถวทั้งหมด (RangeIndex), รายชื่อคอลัมน์พร้อมจำนวนค่าที่ไม่ใช่ null (Non-Null Count) และชนิดข้อมูล (Dtype) ของแต่ละคอลัมน์ รวมถึงขนาดหน่วยความจำที่ใช้ (memory usage)

การตรวจสอบข้อมูลเบื้องต้นแบบนี้เป็นขั้นตอนสำคัญก่อนเริ่มวิเคราะห์ข้อมูลจริง เพื่อให้รู้ว่าข้อมูลมีขนาดเท่าไร มีค่าขาดหายหรือไม่ และแต่ละคอลัมน์เก็บข้อมูลชนิดใด

In [6]:
### Checking each item separately ###

# (1) Number of records
print("Number of records :", df.shape[0])

# (2) Number of columns
print("Number of columns :", df.shape[1])

# (3) Data type of each column
print("\nData type of each column:")
print(df.dtypes)

# (4) Number of non-null values
print("\nNumber of non-null values:")
print(df.notnull().sum())

Number of records : 2205
Number of columns : 27

Data type of each column:
No                     int64
Income                 int64
Kidhome                int64
Teenhome               int64
Recency                int64
MntWines               int64
MntFruits              int64
MntMeatProducts        int64
MntFishProducts        int64
MntSweetProducts       int64
MntGoldProds           int64
NumDealsPurchases      int64
NumWebPurchases        int64
NumCatalogPurchases    int64
NumStorePurchases      int64
NumWebVisitsMonth      int64
AcceptedCmp3           int64
AcceptedCmp4           int64
AcceptedCmp5           int64
AcceptedCmp1           int64
AcceptedCmp2           int64
Complain               int64
Response               int64
Age                    int64
Customer_Days          int64
Marital                  str
Education                str
dtype: object

Number of non-null values:
No                     2205
Income                 2205
Kidhome                2205
Teenhome        

In [7]:
### All information at once ###
df.info()

<class 'pandas.DataFrame'>
RangeIndex: 2205 entries, 0 to 2204
Data columns (total 27 columns):
 #   Column               Non-Null Count  Dtype
---  ------               --------------  -----
 0   No                   2205 non-null   int64
 1   Income               2205 non-null   int64
 2   Kidhome              2205 non-null   int64
 3   Teenhome             2205 non-null   int64
 4   Recency              2205 non-null   int64
 5   MntWines             2205 non-null   int64
 6   MntFruits            2205 non-null   int64
 7   MntMeatProducts      2205 non-null   int64
 8   MntFishProducts      2205 non-null   int64
 9   MntSweetProducts     2205 non-null   int64
 10  MntGoldProds         2205 non-null   int64
 11  NumDealsPurchases    2205 non-null   int64
 12  NumWebPurchases      2205 non-null   int64
 13  NumCatalogPurchases  2205 non-null   int64
 14  NumStorePurchases    2205 non-null   int64
 15  NumWebVisitsMonth    2205 non-null   int64
 16  AcceptedCmp3         2205 non-null 

---

### 5. Describe Numerical Data
Display the descriptive statistics of the numerical columns, including at least: (1) Mean, (2) Standard deviation, (3) Minimum, (4) Maximum

**คำอธิบาย:** ข้อนี้ใช้เมธอด `df.describe()` เพื่อดูค่าสถิติเชิงพรรณนา (descriptive statistics) ของคอลัมน์ที่เป็นตัวเลขทั้งหมดในคราวเดียว โดย default จะแสดง 8 ค่าสถิติ ได้แก่ count, mean, std, min, 25%, 50% (median), 75%, max

เนื่องจากโจทย์ต้องการเฉพาะ mean, std, min, max จึงใช้ `.loc[["mean", "std", "min", "max"]]` ต่อท้าย `.describe()` เพื่อกรองเฉพาะแถวที่มีชื่อ index ตรงกับ 4 ค่าที่ต้องการ (เพราะผลลัพธ์ของ `.describe()` เป็น DataFrame ที่มีชื่อค่าสถิติเป็น index) วิธีนี้ช่วยให้ได้ตารางสรุปที่กระชับ ตรงตามความต้องการ โดยไม่ต้องดูค่าเปอร์เซ็นไทล์ที่ไม่จำเป็น

In [8]:
df.describe()

,No,Income,Kidhome,Teenhome,Recency,MntWines,MntFruits,MntMeatProducts,MntFishProducts,MntSweetProducts,...,NumWebVisitsMonth,AcceptedCmp3,AcceptedCmp4,AcceptedCmp5,AcceptedCmp1,AcceptedCmp2,Complain,Response,Age,Customer_Days
count,2205.000000,2205.000000,2205.000000,2205.000000,2205.000000,2205.000000,2205.000000,2205.000000,2205.000000,2205.000000,...,2205.000000,2205.000000,2205.000000,2205.000000,2205.000000,2205.000000,2205.000000,2205.00000,2205.000000,2205.000000
mean,1103.000000,51622.094785,0.442177,0.506576,49.009070,306.164626,26.403175,165.312018,37.756463,27.128345,...,5.336961,0.073923,0.074376,0.073016,0.064399,0.013605,0.009070,0.15102,51.095692,2512.718367
std,636.672993,20713.063826,0.537132,0.544380,28.932111,337.493839,39.784484,217.784507,54.824635,41.130468,...,2.413535,0.261705,0.262442,0.260222,0.245518,0.115872,0.094827,0.35815,11.705801,202.563647
min,1.000000,1730.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,...,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.00000,24.000000,2159.000000
25%,552.000000,35196.000000,0.000000,0.000000,24.000000,24.000000,2.000000,16.000000,3.000000,1.000000,...,3.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.00000,43.000000,2339.000000
50%,1103.000000,51287.000000,0.000000,0.000000,49.000000,178.000000,8.000000,68.000000,12.000000,8.000000,...,6.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.00000,50.000000,2515.000000
75%,1654.000000,68281.000000,1.000000,1.000000,74.000000,507.000000,33.000000,232.000000,50.000000,34.000000,...,7.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.00000,61.000000,2688.000000
max,2205.000000,113734.000000,2.000000,2.000000,99.000000,1493.000000,199.000000,1725.000000,259.000000,262.000000,...,20.000000,1.000000,1.000000,1.000000,1.000000,1.000000,1.000000,1.00000,80.000000,2858.000000


In [9]:
### Only the required statistics ###
df.describe().loc[["mean", "std", "min", "max"]]

,No,Income,Kidhome,Teenhome,Recency,MntWines,MntFruits,MntMeatProducts,MntFishProducts,MntSweetProducts,...,NumWebVisitsMonth,AcceptedCmp3,AcceptedCmp4,AcceptedCmp5,AcceptedCmp1,AcceptedCmp2,Complain,Response,Age,Customer_Days
mean,1103.000000,51622.094785,0.442177,0.506576,49.009070,306.164626,26.403175,165.312018,37.756463,27.128345,...,5.336961,0.073923,0.074376,0.073016,0.064399,0.013605,0.009070,0.15102,51.095692,2512.718367
std,636.672993,20713.063826,0.537132,0.544380,28.932111,337.493839,39.784484,217.784507,54.824635,41.130468,...,2.413535,0.261705,0.262442,0.260222,0.245518,0.115872,0.094827,0.35815,11.705801,202.563647
min,1.000000,1730.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,...,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.00000,24.000000,2159.000000
max,2205.000000,113734.000000,2.000000,2.000000,99.000000,1493.000000,199.000000,1725.000000,259.000000,262.000000,...,20.000000,1.000000,1.000000,1.000000,1.000000,1.000000,1.000000,1.00000,80.000000,2858.000000


---

### 6. Select One Column
Select and display only column "Income"

**คำอธิบาย:** ข้อนี้แสดงให้เห็นความแตกต่างระหว่างการเลือกคอลัมน์ด้วย single bracket กับ double bracket:
- `df["Income"]` (single bracket) >>> ผลลัพธ์เป็น **Series** คือโครงสร้างข้อมูล 1 มิติ ที่มีเฉพาะค่าของคอลัมน์นั้นและ index กำกับ ไม่มีแนวคิดเรื่อง "คอลัมน์" อีกต่อไป
- `df[["Income"]]` (double bracket) >>> ผลลัพธ์เป็น **DataFrame** คือโครงสร้างข้อมูล 2 มิติ (ยังมีทั้งแถวและคอลัมน์) แม้จะมีข้อมูลแค่คอลัมน์เดียว เพราะเราส่ง list ของชื่อคอลัมน์ `["Income"]` เข้าไปแทนที่จะส่งชื่อคอลัมน์เดี่ยว ๆ

ความแตกต่างนี้สำคัญเวลาที่ต้องนำผลลัพธ์ไปใช้ต่อ เช่น ถ้าต้องการ merge หรือ join ข้อมูลแบบ DataFrame กับ DataFrame มักต้องการ double bracket แต่ถ้าต้องการนำไปคำนวณทางคณิตศาสตร์แบบ element-wise มักใช้ Series ก็เพียงพอ

In [10]:
### As a Series ###
df["Income"]

0       58138
1       46344
2       71613
3       26646
4       58293
        ...  
2200    61223
2201    64014
2202    56981
2203    69245
2204    52869
Name: Income, Length: 2205, dtype: int64

In [11]:
### As a DataFrame (double brackets) ###
df[["Income"]]

,Income
0,58138
1,46344
2,71613
3,26646
4,58293
...,...
2200,61223
2201,64014
2202,56981
2203,69245


---

### 7. Select Multiple Columns
Select and display following columns: "Age", "Income", "Education", "Marital"

**คำอธิบาย:** ข้อนี้ต่อยอดจากข้อ 6 คือเลือกหลายคอลัมน์พร้อมกันโดยใช้ double brackets `df[["Age", "Income", "Education", "Marital"]]` วงเล็บชั้นนอกคือการ index เข้าไปที่ DataFrame ส่วนวงเล็บชั้นในคือ list ของชื่อคอลัมน์ที่ต้องการ

ผลลัพธ์ที่ได้จะเป็น DataFrame ใหม่ (ไม่ใช่ Series) ที่มีเฉพาะ 4 คอลัมน์ตามลำดับที่ระบุไว้ใน list โดยลำดับคอลัมน์ในผลลัพธ์จะเรียงตามลำดับที่เขียนไว้ใน list ไม่ใช่ตามลำดับเดิมในไฟล์ต้นฉบับ วิธีนี้เหมาะกับการดึงเฉพาะคอลัมน์ที่สนใจออกมาดูโดยไม่ต้องแสดงข้อมูลทั้งหมด 27 คอลัมน์

In [12]:
df[["Age", "Income", "Education", "Marital"]]

,Age,Income,Education,Marital
0,63,58138,Graduation,Single
1,66,46344,Graduation,Single
2,55,71613,Graduation,Together
3,36,26646,Graduation,Together
4,39,58293,PhD,Married
...,...,...,...,...
2200,53,61223,Graduation,Married
2201,74,64014,PhD,Together
2202,39,56981,Graduation,Divorced
2203,64,69245,Master,Together


---

### 8. Select Customer Information
Create new DataFrame consisting of "No", "Age", "Income", "Kidhome", "Teenhome", "Marital"

**คำอธิบาย:** ข้อนี้ต้องการสร้าง DataFrame ใหม่ที่มีเฉพาะคอลัมน์ที่เกี่ยวกับข้อมูลลูกค้าโดยตรง โดยเลือกคอลัมน์ "No", "Age", "Income", "Kidhome", "Teenhome", "Marital" ด้วย double brackets เหมือนข้อ 7 แล้วใช้ `.copy()` ต่อท้าย

การใส่ `.copy()` มีความสำคัญมาก เพราะการเลือกคอลัมน์แบบ `df[[...]]` โดยไม่ใส่ `.copy()` จะได้ DataFrame ใหม่ที่ยังอ้างอิง (view) กับข้อมูลเดิมอยู่ในบางกรณี หากนำไปแก้ไขค่าต่อภายหลังอาจเกิดคำเตือน `SettingWithCopyWarning` หรือกระทบข้อมูลต้นฉบับโดยไม่ตั้งใจ การใช้ `.copy()` จึงทำให้มั่นใจได้ว่า `customer_info` เป็น DataFrame อิสระที่แก้ไขได้อย่างปลอดภัยโดยไม่กระทบ `df` ต้นฉบับ

จากนั้นแสดงขนาดของ DataFrame ใหม่ด้วย `.shape` เพื่อตรวจสอบว่าได้จำนวนแถวเท่าเดิม แต่จำนวนคอลัมน์ลดลงเหลือ 6 คอลัมน์ตามที่ต้องการ

In [13]:
customer_info = df[["No", "Age", "Income", "Kidhome", "Teenhome", "Marital"]].copy()

print("Shape :", customer_info.shape)
customer_info

Shape : (2205, 6)


,No,Age,Income,Kidhome,Teenhome,Marital
0,1,63,58138,0,0,Single
1,2,66,46344,1,1,Single
2,3,55,71613,0,0,Together
3,4,36,26646,1,0,Together
4,5,39,58293,1,0,Married
...,...,...,...,...,...,...
2200,2201,53,61223,0,1,Married
2201,2202,74,64014,2,1,Together
2202,2203,39,56981,0,0,Divorced
2203,2204,64,69245,0,1,Together


---

### 9. Select Rows Using .loc[ ]
Use .loc[ ] to select records with index values from 100 to 109 and display only the following columns: "No", "Age", "Income"

**คำอธิบาย:** ข้อนี้ใช้ `.loc[ ]` ซึ่งเป็นการเลือกข้อมูลโดยอ้างอิงจาก "label" ของ index และชื่อคอลัมน์ (ต่างจาก `.iloc[ ]` ที่อ้างอิงตามตำแหน่งตัวเลข) รูปแบบคือ `df.loc[แถวที่ต้องการ, คอลัมน์ที่ต้องการ]`

`df.loc[100:109, ["No", "Age", "Income"]]` หมายถึง:
- เลือกแถวที่มีค่า index ตั้งแต่ 100 ถึง 109 โดยข้อควรระวังคือ `.loc[]` จะ "รวม" ค่าขอบขวาไว้ด้วย (inclusive) ต่างจากการ slice แบบปกติของ Python ดังนั้นจะได้ทั้งหมด 10 แถว (100, 101, ..., 109)
- เลือกเฉพาะคอลัมน์ "No", "Age", "Income" โดยระบุเป็น list ของชื่อคอลัมน์

วิธีนี้เหมาะกับการเลือกข้อมูลเมื่อทราบชื่อคอลัมน์หรือค่า index ที่ต้องการอยู่แล้ว

In [14]:
df.loc[100:109, ["No", "Age", "Income"]]

,No,Age,Income
100,101,60,89058
101,102,41,77298
102,103,29,68126
103,104,57,57288
104,105,29,86037
105,106,40,43974
106,107,58,50785
107,108,50,90765
108,109,42,36550
109,110,53,30753


---

### 10. Select Rows Using .iloc[ ]
Use .iloc[ ] to select records 100–109, columns 1–5

**คำอธิบาย:** ข้อนี้ใช้ `.iloc[ ]` ซึ่งต่างจาก `.loc[ ]` ตรงที่ `.iloc[ ]` อ้างอิงตำแหน่ง (position) แบบตัวเลขล้วน ๆ ไม่สนใจค่า label ของ index หรือชื่อคอลัมน์ โดยรูปแบบคือ `df.iloc[แถวเริ่ม:แถวจบ, คอลัมน์เริ่ม:คอลัมน์จบ]`

`df.iloc[100:110, 1:6]` หมายถึง:
- แถว: ตำแหน่งที่ 100 ถึง 109 (ไม่รวมตำแหน่งที่ 110 เพราะการ slice ด้วยตัวเลขใน iloc ไม่รวมขอบขวา เหมือนการ slice list ทั่วไปของ Python) รวมเป็น 10 แถว
- คอลัมน์: ตำแหน่งที่ 1 ถึง 5 (ไม่รวมตำแหน่งที่ 6) คือคอลัมน์ Income, Kidhome, Teenhome, Recency, MntWines (นับตำแหน่ง 0 คือ No)

ข้อแตกต่างสำคัญระหว่าง `.loc[100:109]` (ข้อ 9) กับ `.iloc[100:110]` (ข้อนี้) คือ `.loc[]` รวมขอบขวา (100 ถึง 109 รวม 109) ในขณะที่ `.iloc[]` ไม่รวมขอบขวา จึงต้องใช้ 110 เพื่อให้ได้แถวสุดท้ายคือตำแหน่ง 109 เช่นกัน

In [15]:
df.iloc[100:110, 1:6]

,Income,Kidhome,Teenhome,Recency,MntWines
100,89058,0,0,18,454
101,77298,0,1,46,425
102,68126,0,0,40,1332
103,57288,0,1,27,527
104,86037,0,0,95,490
105,43974,1,0,19,378
106,50785,1,1,27,64
107,90765,0,0,25,547
108,36550,1,0,74,47
109,30753,1,1,85,12


---

### 11. Filter by Age
Select only customers who are older than 40 years. Display: "No", "Age", "Income", "Response".

**คำอธิบาย:** ข้อนี้ต้องการกรอง (filter) แถวข้อมูลตามเงื่อนไขที่กำหนด โดยใช้ `df.loc[เงื่อนไข, [คอลัมน์ที่ต้องการ]]`
- เงื่อนไข `df["Age"] > 40` จะสร้าง Series ของค่า True/False ที่บอกว่าลูกค้าแต่ละแถวมีอายุมากกว่า 40 ปีหรือไม่
- เมื่อนำ Series นี้ไปใส่ใน `.loc[]` pandas จะเลือกเฉพาะแถวที่มีค่าเป็น True เท่านั้น
- ส่วนที่สองใน `.loc[]` คือ list ของชื่อคอลัมน์ `["No", "Age", "Income", "Response"]` ซึ่งใช้ระบุว่าต้องการแสดงเฉพาะคอลัมน์เหล่านี้

ผลลัพธ์ที่ได้คือ DataFrame ที่มีเฉพาะลูกค้าอายุมากกว่า 40 ปี และแสดงเพียง 4 คอลัมน์ตามที่โจทย์กำหนด ถือเป็นรูปแบบพื้นฐานของการกรองข้อมูลแบบมีเงื่อนไข (conditional filtering) ที่ใช้บ่อยมากในการวิเคราะห์ข้อมูล

In [16]:
df.loc[df["Age"] > 40, ["No", "Age", "Income", "Response"]]

,No,Age,Income,Response
0,1,63,58138,1
1,2,66,46344,0
2,3,55,71613,0
5,6,53,62513,0
6,7,49,55635,0
...,...,...,...,...
2199,2200,46,34421,0
2200,2201,53,61223,0
2201,2202,74,64014,0
2203,2204,64,69245,0


---

### 12. Filter by Income
Select only customers whose Income is greater than 50,000. Display: "No", "Age", "Income".

**คำอธิบาย:** ข้อนี้คล้ายกับข้อ 11 แต่เปลี่ยนเงื่อนไขมาเป็นการกรองจากคอลัมน์ "Income" แทน โดยใช้ `df.loc[df["Income"] > 50000, ["No", "Age", "Income"]]` เพื่อเลือกเฉพาะลูกค้าที่มีรายได้มากกว่า 50,000 และแสดงเฉพาะ 3 คอลัมน์ที่ต้องการ

ข้อนี้ฝึกให้คุ้นเคยกับรูปแบบการกรองข้อมูลตามเงื่อนไขตัวเลข (numeric filtering) ซึ่งเป็นทักษะพื้นฐานที่ใช้บ่อยมากในการวิเคราะห์ข้อมูล เช่น การหากลุ่มลูกค้าที่มีกำลังซื้อสูงเพื่อนำไปทำการตลาดเฉพาะกลุ่ม (targeted marketing)

In [17]:
df.loc[df["Income"] > 50000, ["No", "Age", "Income"]]

,No,Age,Income
0,1,63,58138
2,3,55,71613
4,5,39,58293
5,6,53,62513
6,7,49,55635
...,...,...,...
2200,2201,53,61223
2201,2202,74,64014
2202,2203,39,56981
2203,2204,64,69245


---

### 13. Filter Using Multiple Conditions
Select only customers who meet both conditions: "Age" > 40, "Income" > 50,000

**คำอธิบาย:** ข้อนี้ต้องการกรองข้อมูลด้วย "หลายเงื่อนไขพร้อมกัน" โดยใช้ตัวดำเนินการ AND (`&`) เชื่อมสองเงื่อนไขเข้าด้วยกัน:
- `df["Age"] > 40` >>> อายุมากกว่า 40 ปี
- `df["Income"] > 50000` >>> รายได้มากกว่า 50,000

ใช้รูปแบบ `df[(เงื่อนไข1) & (เงื่อนไข2)]` โดยต้องครอบแต่ละเงื่อนไขด้วยวงเล็บเสมอ เพราะตัวดำเนินการ `&` ใน pandas มีลำดับความสำคัญ (operator precedence) สูงกว่าตัวเปรียบเทียบ (`>`, `<`, `==`) หากไม่ใส่วงเล็บจะเกิด error ผลลัพธ์ที่ได้จะมีเฉพาะแถวที่ผ่าน "ทั้งสองเงื่อนไข" พร้อมกันเท่านั้น (แคบกว่าการกรองด้วยเงื่อนไขเดียว)

In [18]:
df[(df["Age"] > 40) & (df["Income"] > 50000)]

,No,Income,Kidhome,Teenhome,Recency,MntWines,MntFruits,MntMeatProducts,MntFishProducts,MntSweetProducts,...,AcceptedCmp4,AcceptedCmp5,AcceptedCmp1,AcceptedCmp2,Complain,Response,Age,Customer_Days,Marital,Education
0,1,58138,0,0,58,635,88,546,172,88,...,0,0,0,0,0,1,63,2822,Single,Graduation
2,3,71613,0,0,26,426,49,127,111,21,...,0,0,0,0,0,0,55,2471,Together,Graduation
5,6,62513,0,1,16,520,42,98,0,42,...,0,0,0,0,0,0,53,2452,Together,Master
6,7,55635,0,1,34,235,65,164,50,49,...,0,0,0,0,0,0,49,2752,Divorced,Graduation
11,12,63033,0,0,82,194,61,480,225,112,...,0,0,0,0,0,0,61,2385,Divorced,Graduation
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
2194,2195,57967,0,1,39,229,7,137,4,0,...,0,0,0,0,0,0,58,2642,Single,Graduation
2200,2201,61223,0,1,46,709,43,182,42,118,...,0,0,0,0,0,0,53,2540,Married,Graduation
2201,2202,64014,2,1,56,406,0,30,0,0,...,0,0,1,0,0,0,74,2178,Together,PhD
2203,2204,69245,0,1,8,428,30,214,80,30,...,0,0,0,0,0,0,64,2315,Together,Master


---

### 14. Filter Using an OR Condition
Select only customers who meet either of the following conditions: Kidhome > 0 or Teenhome > 0. Display: `No` `Kidhome` `Teenhome`.

**คำอธิบาย:** ข้อนี้ต่างจากข้อ 13 ตรงที่ใช้เงื่อนไข OR (`|`) แทน AND (`&`) หมายความว่าแถวจะถูกเลือกก็ต่อเมื่อ "อย่างน้อยหนึ่งเงื่อนไข" เป็นจริง:
- `df["Kidhome"] > 0` >>> มีลูกเล็กอยู่ในบ้าน
- `df["Teenhome"] > 0` >>> มีลูกวัยรุ่นอยู่ในบ้าน

ใช้ `df.loc[(เงื่อนไข1) | (เงื่อนไข2), [คอลัมน์ที่ต้องการ]]` เพื่อกรองลูกค้าที่มีลูกอย่างน้อยหนึ่งประเภท (ไม่ว่าจะเป็นเด็กเล็กหรือวัยรุ่น หรือทั้งสองอย่าง) แล้วแสดงเฉพาะคอลัมน์ "No", "Kidhome", "Teenhome" ผลลัพธ์ที่ได้จะมีจำนวนแถวมากกว่าการใช้ AND เพราะเงื่อนไข OR หลวมกว่า (ต้องการแค่เงื่อนไขเดียวเป็นจริงก็เพียงพอ)

In [19]:
df.loc[(df["Kidhome"] > 0) | (df["Teenhome"] > 0), ["No", "Kidhome", "Teenhome"]]

,No,Kidhome,Teenhome
1,2,1,1
3,4,1,0
4,5,1,0
5,6,0,1
6,7,0,1
...,...,...,...
2199,2200,1,0
2200,2201,0,1
2201,2202,2,1
2203,2204,0,1


---

### 15. Filter Campaign Responses
Select only customers who accepted the offer in both Campaign 1 and Campaign 2. Conditions: `AcceptedCmp1 == 1` and `AcceptedCmp2 == 1`
Display: `No` `AcceptedCmp1` `AcceptedCmp2` `Response`

**คำอธิบาย:** ข้อนี้คล้ายกับข้อ 13 คือใช้เงื่อนไข AND (`&`) เพื่อกรองแถวที่ตรงกับทั้งสองเงื่อนไขพร้อมกัน แต่เปลี่ยนบริบทมาเป็นการวิเคราะห์แคมเปญการตลาด:
- `df["AcceptedCmp1"] == 1` >>> ลูกค้าที่ตอบรับข้อเสนอในแคมเปญที่ 1
- `df["AcceptedCmp2"] == 1` >>> ลูกค้าที่ตอบรับข้อเสนอในแคมเปญที่ 2

ใช้ `df.loc[เงื่อนไข, [คอลัมน์ที่ต้องการ]]` เพื่อกรองเฉพาะลูกค้าที่ตอบรับทั้งสองแคมเปญ (ตัวเลขน้อยเพราะเป็นเงื่อนไขที่เข้มงวด) แล้วแสดงเฉพาะคอลัมน์ "No", "AcceptedCmp1", "AcceptedCmp2", "Response" เพื่อให้เห็นด้วยว่าลูกค้ากลุ่มนี้ตอบรับแคมเปญล่าสุด (Response) ด้วยหรือไม่ ซึ่งเป็นข้อมูลที่มีประโยชน์ในการหากลุ่มลูกค้าที่มีแนวโน้มตอบรับข้อเสนอสูง (high-value / repeat responders)

In [20]:
df.loc[(df["AcceptedCmp1"] == 1) & (df["AcceptedCmp2"] == 1), ["No", "AcceptedCmp1", "AcceptedCmp2", "Response"]]

,No,AcceptedCmp1,AcceptedCmp2,Response
238,239,1,1,1
400,401,1,1,0
413,414,1,1,1
542,543,1,1,1
558,559,1,1,1
784,785,1,1,1
804,805,1,1,1
970,971,1,1,1
1067,1068,1,1,1
1287,1288,1,1,1


---

### 16. Set the Index
Set `No` as the index of the DataFrame. Then display the first 10 records.

**คำอธิบาย:** ข้อนี้ต้องการเปลี่ยน index (แถวอ้างอิง) ของ DataFrame จากค่าตัวเลขเริ่มต้น (0, 1, 2, ...) ให้เป็นคอลัมน์ "No" แทน โดยใช้เมธอด `df.set_index("No")`

เมื่อกำหนดให้ "No" เป็น index แล้ว คอลัมน์ "No" จะย้ายจากการเป็นคอลัมน์ข้อมูลไปเป็นป้ายกำกับแถว (row label) แทน ทำให้สามารถอ้างอิงแถวด้วยหมายเลขลูกค้าได้โดยตรง เช่น `df_indexed.loc[5]` จะเรียกข้อมูลของลูกค้าหมายเลข 5 ได้ทันที ผลลัพธ์ถูกเก็บไว้ในตัวแปรใหม่ `df_indexed` แล้วแสดง 10 แถวแรกด้วย `.head(10)` เพื่อยืนยันว่า index เปลี่ยนไปตามที่ต้องการ

In [21]:
df_indexed = df.set_index("No")
df_indexed.head(10)

,Income,Kidhome,Teenhome,Recency,MntWines,MntFruits,MntMeatProducts,MntFishProducts,MntSweetProducts,MntGoldProds,...,AcceptedCmp4,AcceptedCmp5,AcceptedCmp1,AcceptedCmp2,Complain,Response,Age,Customer_Days,Marital,Education
No,,,,,,,,,,,,,,,,,,,,,
1,58138,0,0,58,635,88,546,172,88,88,...,0,0,0,0,0,1,63,2822,Single,Graduation
2,46344,1,1,38,11,1,6,2,1,6,...,0,0,0,0,0,0,66,2272,Single,Graduation
3,71613,0,0,26,426,49,127,111,21,42,...,0,0,0,0,0,0,55,2471,Together,Graduation
4,26646,1,0,26,11,4,20,10,3,5,...,0,0,0,0,0,0,36,2298,Together,Graduation
5,58293,1,0,94,173,43,118,46,27,15,...,0,0,0,0,0,0,39,2320,Married,PhD
6,62513,0,1,16,520,42,98,0,42,14,...,0,0,0,0,0,0,53,2452,Together,Master
7,55635,0,1,34,235,65,164,50,49,27,...,0,0,0,0,0,0,49,2752,Divorced,Graduation
8,33454,1,0,32,76,10,56,3,1,23,...,0,0,0,0,0,0,35,2576,Married,PhD
9,30351,1,0,19,14,0,24,3,3,2,...,0,0,0,0,0,1,46,2547,Together,PhD


---

### 17. Reset the Index
Using the DataFrame from Exercise 16, restore `No` as a column.

**คำอธิบาย:** ข้อนี้เป็นการย้อนกลับการกระทำของข้อ 16 โดยใช้เมธอด `.reset_index()` กับ `df_indexed` (ที่เคยตั้ง "No" เป็น index ไปแล้ว) เพื่อดึงคอลัมน์ index กลับมาเป็นคอลัมน์ปกติในตาราง และแทนที่ด้วย index แบบตัวเลขเริ่มต้น (RangeIndex: 0, 1, 2, ...) ตามค่าเริ่มต้นของ pandas

ผลลัพธ์คือ DataFrame ที่มีคอลัมน์ "No" กลับมาอยู่ในตำแหน่งคอลัมน์แรกเหมือนเดิมก่อนที่จะ set_index การสลับไปมาระหว่าง set_index และ reset_index มีประโยชน์เมื่อบางครั้งต้องการใช้ index สำหรับการค้นหาข้อมูลด้วย `.loc[]` แต่บางครั้งก็ต้องการคอลัมน์นั้นกลับมาสำหรับการคำนวณหรือส่งออกข้อมูล

In [22]:
df_indexed = df_indexed.reset_index()
df_indexed.head(10)

,No,Income,Kidhome,Teenhome,Recency,MntWines,MntFruits,MntMeatProducts,MntFishProducts,MntSweetProducts,...,AcceptedCmp4,AcceptedCmp5,AcceptedCmp1,AcceptedCmp2,Complain,Response,Age,Customer_Days,Marital,Education
0,1,58138,0,0,58,635,88,546,172,88,...,0,0,0,0,0,1,63,2822,Single,Graduation
1,2,46344,1,1,38,11,1,6,2,1,...,0,0,0,0,0,0,66,2272,Single,Graduation
2,3,71613,0,0,26,426,49,127,111,21,...,0,0,0,0,0,0,55,2471,Together,Graduation
3,4,26646,1,0,26,11,4,20,10,3,...,0,0,0,0,0,0,36,2298,Together,Graduation
4,5,58293,1,0,94,173,43,118,46,27,...,0,0,0,0,0,0,39,2320,Married,PhD
5,6,62513,0,1,16,520,42,98,0,42,...,0,0,0,0,0,0,53,2452,Together,Master
6,7,55635,0,1,34,235,65,164,50,49,...,0,0,0,0,0,0,49,2752,Divorced,Graduation
7,8,33454,1,0,32,76,10,56,3,1,...,0,0,0,0,0,0,35,2576,Married,PhD
8,9,30351,1,0,19,14,0,24,3,3,...,0,0,0,0,0,1,46,2547,Together,PhD
9,10,5648,1,1,68,28,0,6,1,1,...,0,0,0,0,0,0,70,2267,Together,PhD


---

### 18. Check Missing Values
Check the number of missing values in each column and identify the column with the most missing values.
** Try removing some data from excel and reload again.

Note: `ifood.xlsx` itself has no missing values, so a modified copy (`ifood_missing.xlsx`, some cells cleared in Excel) is loaded here to demonstrate missing-value detection.

**คำอธิบาย:** ข้อนี้ต้องการตรวจสอบว่าแต่ละคอลัมน์มีค่าที่ขาดหายไป (missing values หรือ NaN) กี่ค่า โดยใช้ `df.isnull().sum()` ซึ่ง `.isnull()` จะคืนค่า True/False สำหรับทุกช่องข้อมูลว่าใช่ค่าที่ขาดหายหรือไม่ แล้ว `.sum()` จะรวมจำนวน True (นับ True เป็น 1) ของแต่ละคอลัมน์

เนื่องจากไฟล์ `ifood.xlsx` ต้นฉบับไม่มีค่าขาดหายเลย (ผลลัพธ์เป็น 0 ทุกคอลัมน์) จึงทำตามคำแนะนำในโจทย์ที่บอกให้ "ลองลบข้อมูลบางส่วนออกจาก Excel แล้วโหลดใหม่" โดยสร้างไฟล์ทดลอง `ifood_missing.xlsx` ที่ลบค่าออกไปบางส่วนแบบสุ่ม (โดยเฉพาะคอลัมน์ Income ที่ลบออกมากที่สุด รองลงมาคือ Age และ Marital) แล้วโหลดไฟล์นี้ขึ้นมาตรวจสอบใหม่ด้วย `.isnull().sum()` อีกครั้ง

จากนั้นใช้ `.idxmax()` เพื่อหาว่าคอลัมน์ใดมีจำนวนค่าขาดหายมากที่สุด ซึ่งในกรณีนี้คือคอลัมน์ "Income"

In [23]:
### Check missing values in the original file ###
print("Missing values in ifood.xlsx:")
print(df.isnull().sum())

Missing values in ifood.xlsx:
No                     0
Income                 0
Kidhome                0
Teenhome               0
Recency                0
MntWines               0
MntFruits              0
MntMeatProducts        0
MntFishProducts        0
MntSweetProducts       0
MntGoldProds           0
NumDealsPurchases      0
NumWebPurchases        0
NumCatalogPurchases    0
NumStorePurchases      0
NumWebVisitsMonth      0
AcceptedCmp3           0
AcceptedCmp4           0
AcceptedCmp5           0
AcceptedCmp1           0
AcceptedCmp2           0
Complain               0
Response               0
Age                    0
Customer_Days          0
Marital                0
Education              0
dtype: int64


In [24]:
### After removing some data in Excel and reloading ###
df_missing = pd.read_excel("ifood_missing.xlsx")

missing_counts = df_missing.isnull().sum()
print("Missing values per column:")
print(missing_counts[missing_counts > 0])

print("\nColumn with the most missing values:", missing_counts.idxmax(),
      f"({missing_counts.max()} missing)")

Missing values per column:
Income     24
Age         5
Marital     3
dtype: int64

Column with the most missing values: Income (24 missing)


---

### 19. Remove Missing Records
Remove records containing missing values and compare the number of records:
* Before removal
* After removal
* Number of records removed

**คำอธิบาย:** ข้อนี้ต้องการลบแถว (record) ที่มีค่าขาดหาย (missing value) ในคอลัมน์ใดก็ตามออกจากข้อมูล โดยใช้เมธอด `df_missing.dropna()` ซึ่ง default จะลบทุกแถวที่มีค่า NaN อย่างน้อย 1 ค่าในแถวนั้น

ขั้นตอนคือ:
1. บันทึกจำนวนแถวก่อนลบไว้ในตัวแปร `before_count` ด้วย `.shape[0]`
2. เรียก `.dropna()` เพื่อสร้าง DataFrame ใหม่ (`df_dropped`) ที่ไม่มีแถวซึ่งมีค่าขาดหายเลย
3. นับจำนวนแถวหลังลบ (`after_count`) แล้วคำนวณผลต่างเพื่อดูว่ามีกี่แถวที่ถูกลบออกไป (`removed_count`)

วิธีนี้ (listwise deletion) เป็นวิธีที่ง่ายที่สุดในการจัดการค่าขาดหาย แต่ข้อเสียคือหากมีค่าขาดหายกระจายอยู่ในหลายคอลัมน์ อาจทำให้สูญเสียข้อมูลจำนวนมาก จึงควรเปรียบเทียบจำนวนแถวก่อน-หลังเสมอเพื่อประเมินผลกระทบ

In [25]:
before_count = df_missing.shape[0]

df_dropped = df_missing.dropna()

after_count = df_dropped.shape[0]
removed_count = before_count - after_count

print("Before removal :", before_count)
print("After removal  :", after_count)
print("Records removed:", removed_count)

Before removal : 2205
After removal  : 2173
Records removed: 32


---

### 20. Fill Missing Values
Create a new DataFrame and replace missing values in `Income` with the mean value of `Income`.

**คำอธิบาย:** ข้อนี้ต้องการเติมค่าที่ขาดหาย (missing values) ในคอลัมน์ "Income" แทนที่จะลบแถวนั้นทิ้งไปเลยเหมือนข้อ 19 โดยมีขั้นตอนดังนี้:
1. คัดลอก `df_missing` มาเป็น `df_filled` ด้วย `.copy()` เพื่อไม่ให้กระทบข้อมูลต้นฉบับ
2. คำนวณค่าเฉลี่ย (mean) ของคอลัมน์ Income จากค่าที่ไม่ขาดหาย ด้วย `df_filled["Income"].mean()` (pandas จะข้าม NaN โดยอัตโนมัติเมื่อคำนวณ mean)
3. ใช้ `.fillna(income_mean)` เพื่อแทนที่ค่า NaN ทุกตำแหน่งในคอลัมน์ Income ด้วยค่าเฉลี่ยที่คำนวณได้
4. ตรวจสอบผลลัพธ์ว่าไม่มีค่า NaN เหลืออยู่ในคอลัมน์ Income แล้ว ด้วย `.isnull().sum()`

การเติมค่าด้วยค่าเฉลี่ย (mean imputation) เป็นวิธีพื้นฐานที่ช่วยรักษาจำนวนแถวข้อมูลไว้ครบ แทนที่จะสูญเสียข้อมูลจากการลบแถวทิ้ง ซึ่งเหมาะกับกรณีที่สัดส่วนข้อมูลขาดหายไม่มากและคอลัมน์นั้นมีการกระจายตัวไม่เบ้มากนัก

In [26]:
df_filled = df_missing.copy()

income_mean = df_filled["Income"].mean()
df_filled["Income"] = df_filled["Income"].fillna(income_mean)

print("Mean Income used to fill:", income_mean)
print("Missing values in Income after filling:", df_filled["Income"].isnull().sum())
df_filled

Mean Income used to fill: 51632.53645116919
Missing values in Income after filling: 0


,No,Income,Kidhome,Teenhome,Recency,MntWines,MntFruits,MntMeatProducts,MntFishProducts,MntSweetProducts,...,AcceptedCmp4,AcceptedCmp5,AcceptedCmp1,AcceptedCmp2,Complain,Response,Age,Customer_Days,Marital,Education
0,1,58138.0,0,0,58,635,88,546,172,88,...,0,0,0,0,0,1,63.0,2822,Single,Graduation
1,2,46344.0,1,1,38,11,1,6,2,1,...,0,0,0,0,0,0,66.0,2272,Single,Graduation
2,3,71613.0,0,0,26,426,49,127,111,21,...,0,0,0,0,0,0,55.0,2471,Together,Graduation
3,4,26646.0,1,0,26,11,4,20,10,3,...,0,0,0,0,0,0,36.0,2298,Together,Graduation
4,5,58293.0,1,0,94,173,43,118,46,27,...,0,0,0,0,0,0,39.0,2320,Married,PhD
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
2200,2201,61223.0,0,1,46,709,43,182,42,118,...,0,0,0,0,0,0,53.0,2540,Married,Graduation
2201,2202,64014.0,2,1,56,406,0,30,0,0,...,0,0,1,0,0,0,74.0,2178,Together,PhD
2202,2203,56981.0,0,0,91,908,48,217,32,12,...,1,0,0,0,0,0,39.0,2314,Divorced,Graduation
2203,2204,69245.0,0,1,8,428,30,214,80,30,...,0,0,0,0,0,0,64.0,2315,Together,Master


---

### 21. Rename Columns
Rename the following columns: "MntWines" >>> "Wine_Spending", "MntMeatProducts" >>> "Meat_Spending", "MntFishProducts" >>> "Fish_Spending"

**คำอธิบาย:** ข้อนี้ต้องการเปลี่ยนชื่อคอลัมน์ให้สื่อความหมายชัดเจนขึ้น โดยใช้เมธอด `df.rename(columns={...})` ซึ่งรับพารามิเตอร์เป็น dictionary ที่ระบุคู่ "ชื่อเดิม: ชื่อใหม่" ได้แก่:
- "MntWines" >>> "Wine_Spending"
- "MntMeatProducts" >>> "Meat_Spending"
- "MntFishProducts" >>> "Fish_Spending"

ผลลัพธ์จะถูกเก็บไว้ในตัวแปรใหม่ชื่อ `df_renamed` เพื่อไม่ให้กระทบกับ DataFrame เดิม (`df`) จากนั้นตรวจสอบผลลัพธ์ด้วย `df_renamed.columns` เพื่อยืนยันว่าชื่อคอลัมน์ถูกเปลี่ยนตามที่ต้องการ วิธีนี้ช่วยให้ชื่อคอลัมน์อ่านเข้าใจง่ายขึ้น โดยเฉพาะเมื่อจะนำไปแสดงในรายงานหรือกราฟให้ผู้อื่นดู

In [27]:
df_renamed = df.rename(columns={
    "MntWines": "Wine_Spending",
    "MntMeatProducts": "Meat_Spending",
    "MntFishProducts": "Fish_Spending",
})

df_renamed.columns

Index(['No', 'Income', 'Kidhome', 'Teenhome', 'Recency', 'Wine_Spending',
       'MntFruits', 'Meat_Spending', 'Fish_Spending', 'MntSweetProducts',
       'MntGoldProds', 'NumDealsPurchases', 'NumWebPurchases',
       'NumCatalogPurchases', 'NumStorePurchases', 'NumWebVisitsMonth',
       'AcceptedCmp3', 'AcceptedCmp4', 'AcceptedCmp5', 'AcceptedCmp1',
       'AcceptedCmp2', 'Complain', 'Response', 'Age', 'Customer_Days',
       'Marital', 'Education'],
      dtype='str')

---

### 22. Drop Columns
Create a new DataFrame by removing unnecessary columns, such as: "No", "Customer_Days"

**คำอธิบาย:** ข้อนี้ต้องการลบคอลัมน์ที่ไม่จำเป็นออกจาก DataFrame โดยใช้เมธอด `df.drop(columns=[...])` ระบุรายชื่อคอลัมน์ที่ต้องการลบเป็น list คือ "No" (หมายเลขลำดับ ไม่ได้ใช้ในการวิเคราะห์) และ "Customer_Days" (ในตัวอย่างนี้)

ผลลัพธ์จะถูกเก็บไว้เป็น DataFrame ใหม่ชื่อ `df_dropped_cols` โดย `df` เดิมจะไม่ถูกแก้ไข (เพราะ `.drop()` โดย default จะคืนค่า DataFrame ใหม่ ไม่ได้แก้ไข in-place เว้นแต่จะระบุ `inplace=True`) การลบคอลัมน์ที่ไม่จำเป็นช่วยลดขนาดข้อมูลและทำให้ DataFrame ดูกระชับ เหมาะสำหรับขั้นตอนก่อนนำข้อมูลไปวิเคราะห์หรือสร้างโมเดลต่อ

In [28]:
df_dropped_cols = df.drop(columns=["No", "Customer_Days"])
df_dropped_cols.head()

,Income,Kidhome,Teenhome,Recency,MntWines,MntFruits,MntMeatProducts,MntFishProducts,MntSweetProducts,MntGoldProds,...,AcceptedCmp3,AcceptedCmp4,AcceptedCmp5,AcceptedCmp1,AcceptedCmp2,Complain,Response,Age,Marital,Education
0,58138,0,0,58,635,88,546,172,88,88,...,0,0,0,0,0,0,1,63,Single,Graduation
1,46344,1,1,38,11,1,6,2,1,6,...,0,0,0,0,0,0,0,66,Single,Graduation
2,71613,0,0,26,426,49,127,111,21,42,...,0,0,0,0,0,0,0,55,Together,Graduation
3,26646,1,0,26,11,4,20,10,3,5,...,0,0,0,0,0,0,0,36,Together,Graduation
4,58293,1,0,94,173,43,118,46,27,15,...,0,0,0,0,0,0,0,39,Married,PhD


---

### 23. Replace Values
Inspect the values in the "Education" column and then replace: Graduation >>> Graduate, PhD >>> Doctorate.

**คำอธิบาย:** ข้อนี้ต้องการเปลี่ยนค่าข้อความ (string) บางค่าในคอลัมน์ "Education" ให้เป็นชื่อที่สื่อความหมายชัดเจนขึ้น ขั้นตอนมีดังนี้:
1. ตรวจสอบค่าที่มีอยู่จริงก่อนด้วย `df["Education"].unique()` ซึ่งพบว่ามี 5 ค่า ได้แก่ Graduation, PhD, Master, Basic, 2nCycle
2. ใช้เมธอด `.replace()` พร้อมกับ dictionary ที่ระบุคู่ค่าเดิม >>> ค่าใหม่ คือ `{"Graduation": "Graduate", "PhD": "Doctorate"}` เพื่อแทนที่เฉพาะสองค่านี้ ส่วนค่าอื่น (Master, Basic, 2nCycle) จะไม่ถูกเปลี่ยนแปลง
3. ตรวจสอบผลลัพธ์อีกครั้งด้วย `.unique()` เพื่อยืนยันว่าค่าถูกแทนที่ถูกต้อง

การแทนที่ค่าลักษณะนี้มักใช้เมื่อต้องการทำให้ชื่อระดับการศึกษาอ่านเข้าใจง่ายขึ้นหรือให้สอดคล้องกับมาตรฐานที่ต้องการนำไปใช้ต่อ

In [29]:
### Inspect current values ###
df["Education"].unique()

<StringArray>
['Graduation', 'PhD', 'Master', 'Basic', '2nCycle']
Length: 5, dtype: str

In [30]:
### Replace values ###
df["Education"] = df["Education"].replace({
    "Graduation": "Graduate",
    "PhD": "Doctorate",
})

df["Education"].unique()

<StringArray>
['Graduate', 'Doctorate', 'Master', 'Basic', '2nCycle']
Length: 5, dtype: str

---

### 24. Detect Duplicates
Check whether the dataset contains duplicate records. Report: (A) Number of duplicate records, (B) Duplicate records found

Note: `df` has no duplicate records, so a few duplicate rows are appended here (`df_dup`) to demonstrate duplicate detection.

**คำอธิบาย:** ข้อนี้ต้องการตรวจสอบว่าข้อมูลมีแถวที่ซ้ำกันทุกคอลัมน์หรือไม่ โดยใช้เมธอด `df.duplicated()` ซึ่งจะคืนค่า True/False สำหรับแต่ละแถว (True หมายถึงแถวนั้นซ้ำกับแถวก่อนหน้า) แล้วนำมา `.sum()` เพื่อนับจำนวนแถวที่ซ้ำทั้งหมด

เนื่องจากข้อมูลต้นฉบับ `df` ไม่มีแถวซ้ำอยู่แล้ว (ผลลัพธ์เป็น 0) จึงจำลองสถานการณ์โดยสร้าง `df_dup` ที่คัดลอกแถวแรก ๆ (index 0, 1, 2) มาต่อท้ายข้อมูลเดิมด้วย `pd.concat()` เพื่อให้เห็นตัวอย่างการตรวจจับข้อมูลซ้ำจริง ๆ จากนั้น:
- (A) นับจำนวนแถวซ้ำด้วย `.duplicated().sum()`
- (B) แสดงแถวที่ซ้ำทั้งหมดด้วย `.duplicated(keep=False)` ซึ่งจะทำเครื่องหมาย True ทั้งแถวต้นฉบับและแถวที่คัดลอกมา (ต่างจากค่า default ที่จะมองว่ามีแค่แถวหลังเป็นแถวซ้ำ)

In [31]:
### Check duplicates in the original data ###
print("Number of duplicate records in df:", df.duplicated().sum())

Number of duplicate records in df: 0


In [32]:
### Create a copy with some duplicate rows to demonstrate detection ###
df_dup = pd.concat([df, df.iloc[[0, 1, 2]]], ignore_index=True)

# (A) Number of duplicate records
print("(A) Number of duplicate records:", df_dup.duplicated().sum())

# (B) Duplicate records found
print("\n(B) Duplicate records found:")
df_dup[df_dup.duplicated(keep=False)]

(A) Number of duplicate records: 3

(B) Duplicate records found:


,No,Income,Kidhome,Teenhome,Recency,MntWines,MntFruits,MntMeatProducts,MntFishProducts,MntSweetProducts,...,AcceptedCmp4,AcceptedCmp5,AcceptedCmp1,AcceptedCmp2,Complain,Response,Age,Customer_Days,Marital,Education
0,1,58138,0,0,58,635,88,546,172,88,...,0,0,0,0,0,1,63,2822,Single,Graduate
1,2,46344,1,1,38,11,1,6,2,1,...,0,0,0,0,0,0,66,2272,Single,Graduate
2,3,71613,0,0,26,426,49,127,111,21,...,0,0,0,0,0,0,55,2471,Together,Graduate
2205,1,58138,0,0,58,635,88,546,172,88,...,0,0,0,0,0,1,63,2822,Single,Graduate
2206,2,46344,1,1,38,11,1,6,2,1,...,0,0,0,0,0,0,66,2272,Single,Graduate
2207,3,71613,0,0,26,426,49,127,111,21,...,0,0,0,0,0,0,55,2471,Together,Graduate


---

### 25. Remove Duplicates
Remove duplicate records from the dataset and then check whether any duplicates remain.

**คำอธิบาย:** ข้อนี้ต่อเนื่องจากข้อ 24 โดยนำ `df_dup` (ที่มีแถวซ้ำที่จงใจเพิ่มเข้าไป) มาลบแถวที่ซ้ำกันออก ใช้เมธอด `df_dup.drop_duplicates()` ซึ่งจะเก็บแถวแรกที่พบไว้และลบแถวที่ซ้ำกันในลำดับถัดไปทิ้ง (ค่า default คือ `keep="first"`)

จากนั้นเปรียบเทียบผลลัพธ์ 3 ค่า:
- จำนวนแถวก่อนลบ (`df_dup.shape[0]`)
- จำนวนแถวหลังลบ (`df_dup_removed.shape[0]`)
- ตรวจสอบว่ายังมีข้อมูลซ้ำเหลืออยู่หรือไม่ด้วย `df_dup_removed.duplicated().sum()` ซึ่งควรได้ผลลัพธ์เป็น 0 หากลบสำเร็จ

ขั้นตอนนี้เป็นส่วนสำคัญของการทำความสะอาดข้อมูล (data cleaning) เพื่อให้แน่ใจว่าไม่มีข้อมูลลูกค้าคนเดียวกันถูกนับซ้ำในการวิเคราะห์

In [33]:
df_dup_removed = df_dup.drop_duplicates()

print("Records before removal:", df_dup.shape[0])
print("Records after removal :", df_dup_removed.shape[0])
print("Duplicates remaining  :", df_dup_removed.duplicated().sum())

Records before removal: 2208
Records after removal : 2205
Duplicates remaining  : 0


---

### 26. Change Data Type
Check the data type of "Customer_Days" and then change its data type to an appropriate date-related data type and check the result again.

**คำอธิบาย:** ข้อนี้ต้องการตรวจสอบและแปลงชนิดข้อมูล (data type) ของคอลัมน์ "Customer_Days" ให้เหมาะสมยิ่งขึ้น
1. ก่อนแปลง ใช้ `df["Customer_Days"].dtype` เพื่อดูว่าตอนนี้เก็บเป็นชนิดใด ซึ่งพบว่าเป็น `int64` คือตัวเลขจำนวนเต็มธรรมดา (จำนวนวันที่ลูกค้าเป็นสมาชิก)
2. เนื่องจากค่าของคอลัมน์นี้มีความหมายเป็น "จำนวนวัน" ไม่ใช่ "วันที่" ที่ระบุปฏิทิน ชนิดข้อมูลที่เหมาะสมกว่า int ธรรมดาคือ `timedelta64` (ช่วงเวลา) ไม่ใช่ `datetime64` (วันที่ปฏิทิน) จึงใช้ `pd.to_timedelta(df["Customer_Days"], unit="D")` โดยระบุหน่วยเป็นวัน (Day)
3. หลังแปลงแล้ว ตรวจสอบ dtype อีกครั้งเพื่อยืนยันว่าเปลี่ยนเป็น `timedelta64[ns]` เรียบร้อย และค่าจะแสดงในรูปแบบ เช่น "2822 days" แทนที่จะเป็นตัวเลข 2822 เฉย ๆ

การแปลงชนิดข้อมูลให้ตรงกับความหมายจริงช่วยให้สามารถนำไปคำนวณทางวันที่/เวลาต่อได้ง่ายขึ้น เช่น การบวกกับวันที่สมัครสมาชิกเพื่อหาวันปัจจุบัน

In [34]:
### Check the data type before conversion ###
print("Data type before:", df["Customer_Days"].dtype)
df["Customer_Days"].head()

Data type before: int64


0    2822
1    2272
2    2471
3    2298
4    2320
Name: Customer_Days, dtype: int64

In [35]:
### "Customer_Days" represents a number of days, so convert it to timedelta ###
df["Customer_Days"] = pd.to_timedelta(df["Customer_Days"], unit="D")

print("Data type after:", df["Customer_Days"].dtype)
df["Customer_Days"].head()

Data type after: timedelta64[s]


0   2822 days
1   2272 days
2   2471 days
3   2298 days
4   2320 days
Name: Customer_Days, dtype: timedelta64[s]

---

### 27. Create a Response Label
Create a new column named "Response_Label" based on "Response": 1 >>> Accepted, 0 >>> Not Accepted

**คำอธิบาย:** ข้อนี้ต้องการแปลงค่าตัวเลข 0/1 ในคอลัมน์ Response ให้เป็นข้อความที่อ่านเข้าใจง่ายขึ้น โดยใช้เมธอด `.map()` พร้อมกับ dictionary ที่กำหนดว่าค่าไหนแปลงเป็นอะไร:
- `1` >>> "Accepted" (ตอบรับข้อเสนอ)
- `0` >>> "Not Accepted" (ไม่ตอบรับข้อเสนอ)

`.map()` จะจับคู่ค่าที่อยู่ใน Series (คอลัมน์ Response) กับ key ใน dictionary ทีละค่า แล้วแทนที่ด้วย value ที่ตรงกัน ผลลัพธ์จะถูกเก็บไว้ในคอลัมน์ใหม่ชื่อ "Response_Label" ซึ่งช่วยให้ตีความข้อมูลได้ง่ายขึ้นเมื่อนำไปแสดงผลหรือทำรายงาน โดยไม่ต้องจดจำว่าตัวเลข 0 กับ 1 หมายถึงอะไร

In [36]:
df["Response_Label"] = df["Response"].map({1: "Accepted", 0: "Not Accepted"})

df[["No", "Response", "Response_Label"]]

,No,Response,Response_Label
0,1,1,Accepted
1,2,0,Not Accepted
2,3,0,Not Accepted
3,4,0,Not Accepted
4,5,0,Not Accepted
...,...,...,...
2200,2201,0,Not Accepted
2201,2202,0,Not Accepted
2202,2203,0,Not Accepted
2203,2204,0,Not Accepted


---

### 28. Create a new column named "Income_Level"
Divide customers into three groups: (1) Income < 30,000 >>> Low, (2) 30,000–60,000 >>> Medium, (3) Income > 60,000 >>> High

**คำอธิบาย:** ข้อนี้ต้องการแบ่งกลุ่มลูกค้าตามระดับรายได้ (Income) ออกเป็น 3 กลุ่ม โดยสร้างฟังก์ชัน `income_level()` ที่รับค่ารายได้เข้ามาแล้วตรวจสอบเงื่อนไขทีละขั้น (if-elif-else):
- ถ้า Income < 30,000 >>> คืนค่า "Low"
- ถ้า Income อยู่ระหว่าง 30,000–60,000 (รวมถึง 60,000) >>> คืนค่า "Medium"
- นอกเหนือจากนั้น (มากกว่า 60,000) >>> คืนค่า "High"

จากนั้นใช้ `df["Income"].apply(income_level)` เพื่อนำฟังก์ชันนี้ไปประมวลผลกับค่า Income ของลูกค้าทุกแถว แล้วเก็บผลลัพธ์ไว้ในคอลัมน์ใหม่ชื่อ "Income_Level" วิธีนี้เหมาะกับกรณีที่เงื่อนไขการแบ่งกลุ่มมีหลายขั้นตอนหรือซับซ้อนกว่าการเปรียบเทียบแบบง่าย ๆ

In [37]:
def income_level(income):
    if income < 30000:
        return "Low"
    elif income <= 60000:
        return "Medium"
    else:
        return "High"

df["Income_Level"] = df["Income"].apply(income_level)

df[["No", "Income", "Income_Level"]]

,No,Income,Income_Level
0,1,58138,Medium
1,2,46344,Medium
2,3,71613,High
3,4,26646,Low
4,5,58293,Medium
...,...,...,...
2200,2201,61223,High
2201,2202,64014,High
2202,2203,56981,Medium
2203,2204,69245,High


---

### 29. Customer Analysis 1
Group customers by "Education" and calculate: (1) Number of customers, (2) Average Income, (3) Average Age, (4) Average Wine Spending, (5) Average Web Purchases, (6) Average Store Purchases

**คำอธิบาย:** ข้อนี้ต้องการวิเคราะห์ข้อมูลลูกค้าโดยแบ่งกลุ่มตามระดับการศึกษา (Education) แล้วสรุปค่าสถิติของแต่ละกลุ่ม ใช้เมธอด `df.groupby("Education").agg(...)` ซึ่งเป็นการทำงานแบบ "Split-Apply-Combine":
1. **Split** - แบ่งข้อมูลออกเป็นกลุ่มย่อยตามค่าที่ไม่ซ้ำกันในคอลัมน์ Education (เช่น Graduate, Doctorate, Master, Basic, 2nCycle)
2. **Apply** - คำนวณค่าสถิติที่ต้องการในแต่ละกลุ่ม ได้แก่ จำนวนลูกค้า (`count`) และค่าเฉลี่ย (`mean`) ของ Income, Age, MntWines, NumWebPurchases, NumStorePurchases
3. **Combine** - รวมผลลัพธ์ของแต่ละกลุ่มกลับมาเป็นตารางเดียว โดยแต่ละแถวคือระดับการศึกษาหนึ่งกลุ่ม

การตั้งชื่อในรูปแบบ `Average_Income=("Income", "mean")` คือการใช้ "named aggregation" ของ pandas ทำให้ได้ชื่อคอลัมน์ผลลัพธ์ที่สื่อความหมายชัดเจน แทนที่จะได้ชื่อคอลัมน์เดิมซ้ำกัน

In [38]:
education_summary = df.groupby("Education").agg(
    Number_of_Customers=("No", "count"),
    Average_Income=("Income", "mean"),
    Average_Age=("Age", "mean"),
    Average_Wine_Spending=("MntWines", "mean"),
    Average_Web_Purchases=("NumWebPurchases", "mean"),
    Average_Store_Purchases=("NumStorePurchases", "mean"),
)

education_summary

,Number_of_Customers,Average_Income,Average_Age,Average_Wine_Spending,Average_Web_Purchases,Average_Store_Purchases
Education,,,,,,
2nCycle,198,47625.333333,47.207071,202.757576,3.787879,5.595960
Basic,54,20306.259259,42.537037,7.240741,1.888889,2.851852
Doctorate,476,55220.581933,53.848739,409.573529,4.470588,6.123950
Graduate,1113,51984.008086,50.381851,285.804133,4.123989,5.854447
Master,364,52629.579670,53.063187,333.788462,4.043956,5.901099


---

### 30. Customer Analysis 2
Create a Pivot Table with: Index = Education and Columns = Response and Values = Number of customer

**คำอธิบาย:** ข้อนี้ต้องการสร้าง Pivot Table เพื่อดูภาพรวมความสัมพันธ์ระหว่างระดับการศึกษา (Education) กับการตอบรับแคมเปญ (Response) โดยใช้เมธอด `df.pivot_table()`
- `index="Education"` >>> กำหนดให้ค่าของคอลัมน์ Education เป็นแถว (แกนตั้ง) ของตาราง
- `columns="Response"` >>> กำหนดให้ค่าของ Response (0 หรือ 1) เป็นคอลัมน์ (แกนนอน) ของตาราง
- `values="No"` และ `aggfunc="count"` >>> นับจำนวนลูกค้า (นับจากคอลัมน์ No ซึ่งไม่ซ้ำกัน) ที่ตกอยู่ในแต่ละช่อง (Education × Response)

ผลลัพธ์ที่ได้คือตารางไขว้ (cross-tabulation) ที่บอกว่าลูกค้าแต่ละระดับการศึกษา มีจำนวนกี่คนที่ "ตอบรับ" (Response = 1) และกี่คนที่ "ไม่ตอบรับ" (Response = 0) แคมเปญล่าสุด ทำให้เห็นได้ทันทีว่ากลุ่มการศึกษาใดตอบรับแคมเปญในสัดส่วนที่สูงกว่ากลุ่มอื่น

In [39]:
pivot_table = df.pivot_table(
    index="Education",
    columns="Response",
    values="No",
    aggfunc="count",
)

pivot_table

Response,0,1
Education,,
2nCycle,176,22
Basic,52,2
Doctorate,375,101
Graduate,961,152
Master,308,56
